### ISSUE TITLE
Forbid version-range-based skips that match no versions

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#71096]
I introduced a version-range-based test in `master` with `skip: " - 7.99.99"` but accidentally failed to adjust this range when backporting to `7.x`:

https://github.com/elastic/elasticsearch/blob/fab1fecbea307a22fb1b3c0a583d3dc328e717a7/rest-api-spec/src/main/resources/rest-api-spec/test/nodes.stats/40_store_stats.yml#L3-L4

If this had been in Java the logic would have depended on the `Version.V_8_0_0` constant which doesn't exist in `7.x` so it would have been quickly caught as a compile failure. However there is no such safety net with the YAML tests so it's very easy to forget to adjust these ranges. I think it would be helpful to fail the build if there are any YAML tests that contain a version-range-based skip which does not match any known versions. This feature would have caught the mistake much sooner.

**Comment by mark-vieira** [comment]
Walk me through what happened here. So when this was backported to `7.x` effectively what happened there was this test was never run. Should it have been? What was the "fix" here? Was it to remove the test or modify the version range to something else?

**Comment by DaveCTurner** [comment]
Yes, https://github.com/elastic/elasticsearch/pull/58800 should have said `7.8.99` rather than `7.99.99` so that these tests ran on versions ≥7.9.0. Henning addressed this in 44ed2435a9fdeeab9c53f9d8c0136833f21d1694.

**Comment by DaveCTurner** [comment]
It seems like a fairly common mistake to make :)

```
$ git rev-parse HEAD && git grep '7\.99\.99'
6751969ffae48a7946a9574a8fa1a07650b9c902
modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/110_script_score_boost.yml:      version: " - 7.99.99"
modules/runtime-fields-common/src/yamlRestTest/resources/rest-api-spec/test/runtime_fields/40_runtime_mappings.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.shards/10_basic.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.component_template/10_basic.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/features.get_features/10_basic.yml:      version: " - 7.99.99" # Adjust this after backport
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/features.reset_features/10_basic.yml:      version: " - 7.99.99" # Adjust this after backport
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.put_index_template/10_basic.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.simulate_template/10_basic.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/170_cardinality_metric.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/170_cardinality_metric.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/170_cardinality_metric.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/330_auto_date_histogram.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/350_variable_width_histogram.yml:      version: " - 7.99.99"
rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/350_point_in_time.yml:      version: " - 7.99.99"
x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/10_basic.yml:      version: " - 7.99.99"
x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/constant_keyword/10_basic.yml:      version: " - 7.99.99" # TODO: make it 7.6.99 after backport
x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/data_stream/10_basic.yml:      version: " - 7.99.99"
x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/security/authz/50_data_streams.yml:      version: " - 7.99.99"
x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/security/authz/50_data_streams.yml:      version: " - 7.99.99"
x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/unsigned_long/10_basic.yml:      version: " - 7.99.99"
x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/unsigned_long/60_collapse.yml:      version: " - 7.99.99"
```

**Comment by mark-vieira** [comment]
Gotcha. So I think the check here is to catch a scenario in which a version skip definition will _never_ be applicable for that branch. In the case of a mute folks should use `version: "all"`. This is basically analogous to our testing conventions checks which catch a test case which isn't executed by any currently defined tasks.

**Comment by mark-vieira** [comment]
Question, what's the intention of adding `skip: " - 7.99.99"` to begin with?

**Comment by DaveCTurner** [comment]
> Question, what's the intention of adding skip: " - 7.99.99" to begin with?

When writing a feature on `master` we do this so that the BWC tests can continue to run against all the `7.x` versions while the backport is still in flight, since that might take a nontrivial amount of time. This is preferable to simply disabling all BWC tests because normally the BWC impact of a new feature is rather limited: keeping the rest of the tests running reduces the impact on other developers and makes sure we're not breaking anything else in BWC-land.

Then we merge it to `master`, and only disable the BWC tests when merging the backport to `7.x` (with suitably adjusted skips).

**Comment by mark-vieira** [comment]
>  when merging the backport to 7.x (with suitably adjusted skips).

Right and then we _should_ also bump the skip version in `master` as well so these things run during BWC tests going forward on `master`. Is that generally not standard practice? Or do we only adjust the skip version in the `7.x` backport. IMO the `7.99.99` version definition should only exist in `master` for features that truly only exist in `master` and have not been backported.

**Comment by DaveCTurner** [comment]
> Right and then we should also bump the skip version in `master`

Correct, although I don't see a good way to automatically detect cases where that wasn't done. When I'm backporting something with BWC implications I keep a list of all the version bounds that I had to update in the backport so I know to update them in `master` too. Not watertight, but seems reasonably reliable in practice and at least we're not completely skipping the test in `master` ever.

**Comment by mark-vieira** [comment]
> Correct, although I don't see a good way to automatically detect cases where that wasn't done.

Yeah, the ultimate goal is to make it _unnecessary_. As I understand it if I just search for `7.99.99` there's currently no way to really distinguish between tests for master-only functionality (stuff that wasn't backported) and tests there were temporarily skipped until the backport was completed. Is that correct?

**Comment by DaveCTurner** [comment]
Correct
